x/evm/keeper/state_transition.go

```diff
@@ -358,18 +364,21 @@ func (k *Keeper) ApplyMessageWithConfig(ctx sdk.Context,
 
 	evm := k.NewEVM(ctx, msg, cfg, tracer, stateDB)
 	leftoverGas := msg.GasLimit
+	sender := vm.AccountRef(msg.From)
 	// Allow the tracer captures the tx level events, mainly the gas consumption.
 	vmCfg := evm.Config
 	if vmCfg.Tracer != nil {
+		stateDB.SubBalance(sender.Address(), new(big.Int).Mul(msg.GasPrice, new(big.Int).SetUint64(msg.GasLimit)))
```

Comment:
if the node set config `evm.tracer`, the consensus will also have non nil `Tracer`.

Reply:
ok, I thought only tracing RPC calls would have the Tracer parameter, I'll add a separate parameter to indicate whether it's a debug tracer call